### PR TITLE
[New Generator] Rust API client generator

### PR DESCRIPTION
**Pull request by farcaller**
### PR checklist

- [ ] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ ] Filed the PR against the correct branch: master for non-breaking changes and `3.0.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

Tracking PR for Rust client implementation.

The first commit adds a subset of apis requested in #5905.

### CODE REVIEW

**Comment by farcaller** [comment]
I'll add a few comments on the specifics inline

File: samples/client/petstore/rust/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "petstore-client"
```
Comment (farcaller) [inline samples/client/petstore/rust/Cargo.toml]: these settings would be exposed as flags.

File: samples/client/petstore/rust/examples/client.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+extern crate petstore_client;
```
Comment (farcaller) [inline samples/client/petstore/rust/examples/client.rs]: a sample file, not to be part of generated api

File: samples/client/petstore/rust/src/apis/add_pet_api.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+use hyper;
+use serde_json;
+use futures::{Future, Stream};
+use futures;
+
+use super::models;
+use super::Error;
+
+pub fn add_pet<C: hyper::client::Connect>(
```
Comment (farcaller) [inline samples/client/petstore/rust/src/apis/add_pet_api.rs]: rust naming guidelines dictate `snake_case` in function names

File: samples/client/petstore/rust/src/apis/add_pet_api.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    prefix: &str,
```
Comment (farcaller) [inline samples/client/petstore/rust/src/apis/add_pet_api.rs]: I'm not sure about what's the common pattern for api use, really. This sample client will require two things: a http (or https) prefix as a string and the hyper.rs Client (the web client) plus any arguments.

File: samples/client/petstore/rust/src/apis/add_pet_api.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    cli: &hyper::client::Client<C>,
+    pet: &models::Pet,
+    ) -> Box<Future<Item = (), Error = Error>> {
+    let mut req = hyper::Request::new(
+        hyper::Method::Post,
+        format!("{}/pet", prefix).parse().unwrap());
+    let serialized = serde_json::to_string(pet).unwrap();
```
Comment (farcaller) [inline samples/client/petstore/rust/src/apis/add_pet_api.rs]: no error handling for brevity of the example.

File: samples/client/petstore/rust/src/apis/add_pet_api.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    req.headers_mut().set(hyper::header::ContentType::json());
+    req.headers_mut().set(hyper::header::ContentLength(serialized.len() as u64));
+    req.set_body(serialized);
+
+    Box::new(
+        cli.request(req).and_then(|res| { res.body().concat2() })
+        .map_err(|e| Error::from(e))
+        .and_then(|_| futures::future::ok(()))
```
Comment (farcaller) [inline samples/client/petstore/rust/src/apis/add_pet_api.rs]: the api doesn't dictate the return value so nothing is returned even though the sample server returns a pet instance.

npm server choked up on `{"id":1337,"category":null,"name":"barker","photoUrls":[],"tags":[],"status":null}` here, http://petstore.swagger.io worked fine.

File: samples/client/petstore/rust/src/apis/mod.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+impl From<serde_json::Error> for Error {
+    fn from(e: serde_json::Error) -> Self {
+        return Error::Serde(e)
+    }
+}
+
+use super::models;
+
+mod get_pet_by_id_api;
```
Comment (farcaller) [inline samples/client/petstore/rust/src/apis/mod.rs]: there's a single namespace, so modules (and files) would have to be named differently from the actual functions.

File: samples/client/petstore/rust/src/models/pet.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/// a Pet
+///
+/// A pet for sale in the pet store
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Pet {
+    id: Option<i64>,
+    category: Option<super::Category>,
+    name: String,
+    #[serde(rename = "photoUrls")] photo_urls: Vec<String>,
```
Comment (farcaller) [inline samples/client/petstore/rust/src/models/pet.rs]: again, snake case here, and renamed fields need an annotation.

**Comment by wing328** [comment]
@farcaller thanks for the Rust Petstore sample, I'll start the reverse engineering process tomorrow and let you know if I've any question.

**Comment by farcaller** [comment]
I poked more around the go generated code and brought the example to the similar level of api. The changes, basically, are that there's an APIClient now, that incapsulates access to the API, and a dedicated pet_api.rs that's supposedly implements all the Pet apis. This extension fully builds on the previous example, though, so it would be fine if the generator produced the functional API as in original commit, and I can tackle the object API later myself when I'll have a better understanding of swagger internals.

**Comment by wing328** [comment]
Quick question: would it be possible to group `add_pet_api.rs`, `get_pet_by_id_api.rs`, `update_pet_with_form_api.rs` into a single rs file (e.g. pet_api_func.rs)?

**Comment by farcaller** [comment]
Absolutely. I'll update the PR

**Comment by wing328** [comment]
@farcaller I've submitted the initial version via https://github.com/swagger-api/swagger-codegen/pull/6105

Please comment on the code in that PR and we'll fix the remaining issues one by one.

File: samples/client/petstore/rust/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+[dependencies]
+serde = "*"
+serde_derive = "*"
+serde_yaml = "*"
+serde_json = "*"
+base64 = "*"
+futures = "*"
+hyper = "*"
+url = "*"
```
Comment (lucab) [inline samples/client/petstore/rust/Cargo.toml]: Would you mind annotating which dependency version is the codegen targeting?

File: samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+use futures::{Future, Stream};
+
+use super::models;
+use super::Error;
+
+pub fn get_pet_by_id<C: hyper::client::Connect>(
+    prefix: &str,
+    cli: &hyper::client::Client<C>,
+    pet_id: i64,
+    ) -> Box<Future<Item = models::Pet, Error = Error>> {
```
Comment (lucab) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: (Here and it other generated methods)
Would you consider returning a more complex type? The Item here is losing any additional information about the response (header, status, etc) which I've found useful in some situations (e.g. token authenticated endpoints redirecting to token provider). For reference, reqwest does this with a richer `Response` [struct + trait](https://github.com/seanmonstar/reqwest/blob/8296c5e28e00a09ce45e0e33f8b10a8d15105a90/src/response.rs#L27).

File: samples/client/petstore/rust/Cargo.toml (hunk 9 above)
Comment (farcaller) [inline samples/client/petstore/rust/Cargo.toml]: yep, as soon as I figure those myself :-) so far the generated code doesn't compile

File: samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs (hunk 10 above)
Comment (farcaller) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: ack.

Comment (wing328) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: For other generators (e.g. PHP, Ruby), there's a corresponding "WithHttpInfo" method (e.g. getPetByIdWithHttpInfo) to return a list with the response (if any), HTTP status code and HTTP headers.

Not sure if Rust supports returning a list. If not, we will need to create an ApiResponse object with response, status and headers as properties.

For Rust, we can follow similar approach by creating a corresponding `WithHttpInfo` method or we can default the method (e.g. getPetById) to return the response together with status code and headers.

Comment (farcaller) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: I think we can return response object along with the parsed body. Let's focus on making the current implementation compileable and then we can improve.

Comment (wing328) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: It's also worth sharing other API clients usually come with a corresponding "Async" method (e.g. [addPetAsync](https://github.com/swagger-api/swagger-codegen/blob/master/samples/client/petstore/csharp/SwaggerClient/src/IO.Swagger/Api/PetApi.cs#L572)) for async function call. Not sure how that should be done in Rust but worth thinking ahead with this future requirement.

Comment (farcaller) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: All this code is async-only. I don't think that it makes sense to have a matching sync version.

Comment (wing328) [inline samples/client/petstore/rust/src/apis/get_pet_by_id_api.rs]: Ah ok. Good to know. One less thing to consider then.

**Comment by wing328** [comment]
Merging this (Rust Petstore sample) into master first similar to what we've done for PowerShell and other generators that follow similar approach so as to give credits back to the contributor.

For further discussion on the Rust generator, please continue in https://github.com/swagger-api/swagger-codegen/pull/6105, which is work-in-progress.

**Pull request merged**